### ISSUE TITLE
Non contiguous gradient tensor not checked (causes segmentation fault)

### ISSUE THREAD

**Issue by oliver-batchelor** [taichi-dev/taichi#8443]
Non-contiguous gradient vectors currently cause segmentation faults as shown by example below!

```
import taichi as ti
import torch

@ti.kernel
def double_kernel(x:ti.types.ndarray(dtype=ti.f32, ndim=1), 
           out:ti.types.ndarray(dtype=ti.f32, ndim=1)):
  for i in range(x.shape[0]):
    out[i] = 2*x[i]


if __name__ == '__main__':
    ti.init(log_level=ti.DEBUG, debug=True)

    device = torch.device('cpu')
    params = torch.rand(1000, dtype=torch.float32, device=device)
    out = torch.zeros_like(params)

    double_kernel(params, out)

    out.requires_grad_(True)
    grad = torch.tensor([1.], dtype=torch.float32)
    out.grad = grad.expand(1000) # make a non contiguous tensor

    double_kernel.grad(params, out)
```
